### PR TITLE
Document changed scheduler behaviour in Version 7

### PR DESCRIPTION
**Pull request by andreasohlund**
I've [discovered that the scheduler behaviour was changed in Version 7](https://github.com/Particular/NServiceBus/commit/d96be1169c796b9e34ad01495b12bc1126b405f8#diff-696f13b7392fcbb966e1347c09534cf1R38) so this documents this and provides an upgrade guide entry to guide users.

### CODE REVIEW

**Comment by andreasohlund** [comment]
@danielmarbach adding you as a reviewer FYI since you made that change in v7

File: nservicebus/scheduling/index.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ redirects:
 reviewed: 2020-08-24
 ---
 
-The Scheduler is a lightweight/non-durable API that helps schedule a task that needs to be executed repeatedly based on a specified interval. The scheduling infrastructure leverages the approach of reliable messaging with the NServiceBus core functionality. This allows scheduling to include features such as built in retries and forwarding to the error queue.
+The Scheduler is a lightweight/non-durable API that helps schedule a task that needs to be executed repeatedly based on a specified interval. In order to benefit from NServiceBus features such as built in retries and forwarding to the error queue it's recommended that scheduled tasks only sends messages in order to perform the actual work. One example would be to query the database for orders that needs some action to be taken and emit individual messages for each order that is found.
```
Comment (lailabougria) [inline nservicebus/scheduling/index.md:12]:
```suggestion
The Scheduler is a lightweight/non-durable API that helps schedule a task that needs to be executed repeatedly based on a specified interval. In order to benefit from NServiceBus features such as built-in retries and forwarding to the error queue, it's recommended that scheduled tasks only send messages in order to perform the actual work. One example would be to query the database for orders that need some action to be taken and emit individual messages for each order that is found.
```

File: nservicebus/upgrades/6to7/index.md [hunk 2]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
 
  * [NServiceBus Testing](../testing-6to7.md)
  * [Azure Blob Storage Data Bus](../absdatabus-1to2.md)
+
+## Scheduled tasks
+
+In Version 6 scheduled tasks would be invoked as a separate `System.Task` using `Task.Run`.
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:361]:
```suggestion
In Version 6, scheduled tasks were invoked as a separate `System.Task` using `Task.Run`.
```

File: nservicebus/upgrades/6to7/index.md [hunk 3]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
...
+
+This would mean that:
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:363]:
```suggestion
This means that:
```

File: nservicebus/upgrades/6to7/index.md [hunk 4]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
...
+
+* They wouldn't be able participate in the [transport transaction](/transports/transactions)
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:365]:
```suggestion
* They are not able participate in the [transport transaction](/transports/transactions)
```

File: nservicebus/upgrades/6to7/index.md [hunk 5]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
...
+* They wouldn't be limited in time by the timeout of the transport transaction
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:366]:
```suggestion
* They are not limited in time by the timeout of the transport transaction
```

File: nservicebus/upgrades/6to7/index.md [hunk 6]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
...
+
+In Version 7 tasks are executed as part of the receive pipeline using an `await`.
+
+This means that task now will participate in the [transport transaction](/transports/transactions) and is therefor exposed to the transaction timeout.
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:370]:
```suggestion
This means that the task now participates in the [transport transaction](/transports/transactions) and is therefore exposed to the transaction timeout.
```

File: nservicebus/upgrades/6to7/index.md [hunk 7]
```diff
@@ -355,3 +355,20 @@ This article so far covers the changes between NServiceBus 6 and NServiceBus 7.
...
+
+To get the Version 6 behavior please make sure to execute the task in a `Task.Run` as shown below:
```
Comment (lailabougria) [inline nservicebus/upgrades/6to7/index.md:372]:
```suggestion
To achieve the Version 6 behavior, please make sure to execute the task in a `Task.Run` as shown below:
```

File: nservicebus/scheduling/index.md [hunk 8]
```diff
@@ -18,6 +18,7 @@ Scheduling a task depends on [delayed delivery](/nservicebus/messaging/delayed-d
 The Scheduler leverages the queuing system to trigger scheduled actions. Under heavy load there may be some disparity between the expected time of a scheduled action and execution time due to the delay between timeout messages being generated and processed. On approach to mitigating this behavior is to run the Scheduler in a dedicated endpoint so that appropriate resource allocation can put in place to ensure timely execution of scheduled actions. Or, alternatively, use a difference scheduling technology, for example:
 
  * A [.NET Timer](https://msdn.microsoft.com/en-us/library/system.threading.timer.aspx).
+ * [NServiceBus Sagas](/nservicebus/sagas/)
```
Comment (andreasohlund) [inline nservicebus/scheduling/index.md:21]: We do mention this below as well but I think it should be here as well to make the list of options complete

**Comment by danielmarbach** [comment]
Is this the change https://github.com/Particular/NServiceBus/commit/d96be1169c796b9e34ad01495b12bc1126b405f8 ?

File: Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    class ExecuteScheduledTaskUsingTaskRun
+    {
+        async Task TaskRun(IEndpointInstance endpointInstance, ILog logger)
+        {
+            #region 6to7ExecuteScheduledTaskUsingTaskRun
+            await endpointInstance.ScheduleEvery(
+                    timeSpan: TimeSpan.FromMinutes(5),
+                    task: pipelineContext =>
+                    {
+                        Task.Run(() =>
```
Comment (danielmarbach) [inline Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs:17]: I would strongly discourage using the continuation style. 

It is much better to wrap it in a method or local function, something like


```
_ = DoWork();

static async Task DoWork() {
  try {
     // Do some work here
     logger.Error($"Failed to execute MyTask'", ex);
  } catch (Exception ex) {
     logger.InfoFormat("MyTask taskcompleted");
  }
}
```

File: Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                            return Task.CompletedTask;
+                        })
+                        .ContinueWith(task =>
+                        {
+
+                            if (task.IsFaulted)
+                            {
+                                task.Exception.Handle(ex =>
+                                {
+                                    logger.Error($"Failed to execute MyTask'", ex);
```
Comment (danielmarbach) [inline Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs:29]: Unnecessary interpolation

File: Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs (hunk 9 above)
Comment (danielmarbach) [inline Snippets/Core/Core_7/UpgradeGuides/6to7/Scheduling/ExecuteScheduledTaskUsingTaskRun.cs:17]: Why is `Task.Run` even relevant here? I think what you want to achieve is fire & forget a scheduled task to make it independent from the transport transaction

**Comment by andreasohlund** [comment]
Talking to @danielmarbach I realized that the behaviour was actually changed in v6 so since v5 is no longer documented and supported I could simplify the change and raised https://github.com/Particular/docs.particular.net/pull/5042 to replace this

**Pull request closed without merging**